src/session.ts

```diff
@@ -894,21 +616,353 @@ function addTrack(this: WebPhoneSession, remoteAudioEle, localAudioEle): void {
+        }
+    });
+}
+
+function setHold(session: WebPhoneSession, hold: boolean): Promise<void> {
+    return new Promise((resolve, reject) => {
+        // Just resolve if we are already in correct state
+        if (session.held === hold) {
+            return resolve();
+        }
+
+        const options: SessionInviteOptions = {
+            requestDelegate: {
+                onAccept: async (response): Promise<void> => {
+                    session.held = hold;
+                    const sessionDescriptionHandler = session.sessionDescriptionHandler as SessionDescriptionHandler;
+                    const peerConnection = sessionDescriptionHandler.peerConnection;
+                    const localSdp = peerConnection.localDescription.sdp;
+                    const match = localSdp.match(/a=(sendrecv|sendonly|recvonly|inactive)/);
+                    const direction = match ? match[1] : '';
+                    session.__localHold = response.message.statusCode === 200 && direction === 'sendonly';
+                    (session as any).logger.log('localhold is set to ' + session.__localHold);
+                    enableReceiverTracks(session, !session.held);
+                    enableSenderTracks(session, !session.held && !session.muted);
+                    resolve();
+                },
+                onReject: (): void => {
+                    (session as any).logger.warn('re-invite request was rejected');
+                    enableReceiverTracks(session, !session.held);
+                    enableSenderTracks(session, !session.held && !session.muted);
+                    reject();
+                }
+            }
+        };
+
+        // Session properties used to pass options to the SessionDescriptionHandler:
+        //
+        // 1) Session.sessionDescriptionHandlerOptions
+        //    SDH options for the initial INVITE transaction.
+        //    - Used in all cases when handling the initial INVITE transaction as either UAC or UAS.
+        //    - May be set directly at anytime.
+        //    - May optionally be set via constructor option.
+        //    - May optionally be set via options passed to Inviter.invite() or Invitation.accept().
+        //
+        // 2) Session.sessionDescriptionHandlerOptionsReInvite
+        //    SDH options for re-INVITE transactions.
+        //    - Used in all cases when handling a re-INVITE transaction as either UAC or UAS.
+        //    - May be set directly at anytime.
+        //    - May optionally be set via constructor option.
+        //    - May optionally be set via options passed to Session.invite().
+
+        const sessionDescriptionHandlerOptions =
+            session.sessionDescriptionHandlerOptionsReInvite as SessionDescriptionHandlerOptions;
+        sessionDescriptionHandlerOptions.hold = hold;
+        session.sessionDescriptionHandlerOptionsReInvite = sessionDescriptionHandlerOptions;
+
+        // Send re-INVITE
+        session
+            .invite(options)
```

Comment:
Do we still need to set extraHeaders to add contact? https://github.com/ringcentral/ringcentral-web-phone/pull/344/files#diff-96998a2148aa3d52ad3e80303d0ad88fc43a4afd08966b5330a5657021ae52bbR614